### ISSUE TITLE
32 bit and 64 bit Debian containers have different permissions by default

### ISSUE THREAD

**Issue by voltagex** [docker/docker#20550]
Description of problem:
A 32 bit container created with [mkimage/debootstrap](https://raw.githubusercontent.com/docker/docker/master/contrib/mkimage/debootstrap) has different permissions by default than a 64 bit container created with the same script

`docker version`: 

```
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:15:20 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:15:20 2016
 OS/Arch:      linux/amd64
```

`docker info`:
```
Containers: 21
 Running: 0
 Paused: 0
 Stopped: 21
Images: 13
Server Version: 1.10.1
Storage Driver: devicemapper
 Pool Name: docker-8:50-3939729-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 2.217 GB
 Data Space Total: 107.4 GB
 Data Space Available: 105.2 GB
 Metadata Space Used: 2.953 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.115 (2016-01-25)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.3.0-1-amd64
Operating System: Debian GNU/Linux stretch/sid
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 23.43 GiB
Name: xxx
ID: xxx
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No oom kill disable support

```

`uname -a`: Linux xxx 4.3.0-1-amd64 #1 SMP Debian 4.3.5-1 (2016-02-06) x86_64 GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.): Physical Debian sid host

Steps to Reproduce:
1. Build an x64 debian sid image with the script linked above (note ./debootstrap NOT debootstrap)
```
./debootstrap sid-test sid http://httpredir.debian.org/debian/
# coffee time
tar -cf - -C sid-test . | docker import - test-debian-sid
```
2. Patch the ./debootstrap script with the following hack:
```diff
diff --git a/contrib/mkimage/debootstrap b/contrib/mkimage/debootstrap
index c613d53..3882767 100755
--- a/contrib/mkimage/debootstrap
+++ b/contrib/mkimage/debootstrap
@@ -30,7 +30,7 @@ rootfs_chroot() {
 
 (
 	set -x
-	$DEBOOTSTRAP "${before[@]}" "$suite" "$rootfsDir" "$@"
+	$DEBOOTSTRAP --arch=i386 "${before[@]}" "$suite" "$rootfsDir" "$@"
 )
 
 # now for some Docker-specific tweaks

```

3. Create a 32 bit image with the "new" script
```bash
./debootstrap sid-test-i386 sid http://httpredir.debian.org/debian/
# maybe whisky this time?
tar -cf - -C sid-test-i386 . | docker import - test-debian-sid-i386
```

4. Run the commands from `Actual Results` below in each container. Perhaps drink more whiskey, as building a simple Debian package in a Docker container has now taken 6 hours.

Actual Results:
```bash
# 32 bit
sudo docker run -it test-debian-sid-i386 bash

root@5f6adf619d1f:/# ip
Cannot open netlink socket: Operation not permitted
root@5f6adf619d1f:/# ping 8.8.8.8
socket: Operation not permitted
```

```bash
# 64 bit
$ sudo docker run -it test-debian-sid bash
root@9aeead06db9d:/# ip
Usage: ip [ OPTIONS ] OBJECT { COMMAND | help }
       ip [ -force ] -batch filename
where  OBJECT := { link | address | addrlabel | route | rule | neighbor | ntable |
                   tunnel | tuntap | maddress | mroute | mrule | monitor | xfrm |
                   netns | l2tp | fou | tcp_metrics | token | netconf }
       OPTIONS := { -V[ersion] | -s[tatistics] | -d[etails] | -r[esolve] |
                    -h[uman-readable] | -iec |
                    -f[amily] { inet | inet6 | ipx | dnet | mpls | bridge | link } |
                    -4 | -6 | -I | -D | -B | -0 |
                    -l[oops] { maximum-addr-flush-attempts } | -br[ief] |
                    -o[neline] | -t[imestamp] | -ts[hort] | -b[atch] [filename] |
                    -rc[vbuf] [size] | -n[etns] name | -a[ll] | -c[olor]}
root@9aeead06db9d:/# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=54 time=5.88 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=54 time=6.21 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=54 time=5.99 ms
64 bytes from 8.8.8.8: icmp_seq=4 ttl=54 time=5.99 ms
64 bytes from 8.8.8.8: icmp_seq=5 ttl=54 time=5.93 ms
^C
--- 8.8.8.8 ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 4004ms
rtt min/avg/max/mdev = 5.886/6.004/6.211/0.139 ms
```

Expected Results: similar output from each container - i.e. same permissions to use sockets

Additional info: Running with `--privileged` 'fixes' this issue, but why? I'm treating this as a (very annoying) bug because two containers created the same way should work the same by default.

If you manage to get strace in the container (by magic, that script is very difficult to understand), you'll get permission denied on the first ptrace() call.

**Comment by cpuguy83** [comment]
Does this work by adding `--security-opt seccomp:unconfined` to the 32bit container's startup?

**Comment by voltagex** [comment]
@cpuguy83 yes, that works.

**Comment by cpuguy83** [comment]
ping @justincormack don't know... is this an issue of blocking `personality` syscall?

**Comment by justincormack** [comment]
@cpuguy83 I think it unlikely to be related to `personality` as that is only run outside the container by the debootstrap here. But it does look like it is a different seccomp related issue, will replicate and see what it is.

**Comment by voltagex** [comment]
Hi Justin, thanks for looking into this. Let me know if you need any other info or testing done.

**Comment by justincormack** [comment]
@voltagex I can't replicate this by building my own 32 bit debian from your instructions, `ping` and `ip` work fine for me. But I didn't build on the same host, and the host environment could affect it, maybe you could push your image to docker hub and I could test from that?

**Comment by voltagex** [comment]
I pushed it but I can't see it...

```
voltagex@debian:~$ sudo docker push voltagex/debian-sid-i386
The push refers to a repository [docker.io/voltagex/debian-sid-i386]
0cb85261784c: Pushed
latest: digest: sha256:6388bc5dcf80eac7f43a05a54808d2d3af7bc6b29665253d902667ff9fc07e3e size: 1190
```

**Comment by voltagex** [comment]
Uh oh, I have no issues pulling that image for myself and running it on a new AWS instance (Debian Jessie). Kernel version is 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt11-1

**Comment by voltagex** [comment]
As requested in IRC:

```
voltagex@debian:~$ ldd `which docker`
        linux-vdso.so.1 (0x00007ffc368a5000)
        libapparmor.so.1 => /lib/x86_64-linux-gnu/libapparmor.so.1 (0x00007f49a3e98000)
        libseccomp.so.2 => /lib/x86_64-linux-gnu/libseccomp.so.2 (0x00007f49a3c5a000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f49a3a3c000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f49a3838000)
        libdevmapper.so.1.02.1 => /lib/x86_64-linux-gnu/libdevmapper.so.1.02.1 (0x00007f49a35dc000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f49a3237000)
        /lib64/ld-linux-x86-64.so.2 (0x000055f679572000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f49a302f000)
        libselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007f49a2e0d000)
        libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007f49a2dec000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f49a2ae7000)
        libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007f49a2877000)

```

**Comment by justincormack** [comment]
How did you install that version of docker? From the distro or from docker's packages? Guessing there is some sort of `libseccomp` difference from the official binary builds (which are statically linked), and maybe the blacklist of distros to build with seccomp should be changed?

**Comment by voltagex** [comment]
It's installed from Docker's Debian repos in both cases. 

Where I'm up to now:
On Jessie host - 32 bit is working the same as for you
On Sid host - 32 bit doesn't work, throws same original errors

I thought it might be a difference between seccomp 2.2.1 (Jessie) and 2.2.3 (Sid) but I've just built 2.2.3 for Jessie and installed it - still working.

**Comment by justincormack** [comment]
Thanks, that narrows it down quite a bit.

**Comment by justincormack** [comment]
Ok, I can replicate it on Debian testing.

**Comment by justincormack** [comment]
Ok, I think it is due to the use of the direct (non socketcall) versions of the socket-related syscalls on new glibc+new kernel. I will look at the best solution, probably doing exact match filtering and explicitly allowing socketcall.

**Comment by justincormack** [comment]
Also there is about to be a new release of libseccomp which will also resolve this, and which we will also want to upgrade to, for more seccomp support for other architectures. I will test if we want to make changes anyway to support this, seperate for the upgrade,

So the current situation is you should disable seccomp on a host kernel 4.3 or later if running 32 bit binaries that access the network.

**Comment by justincormack** [comment]
The new libseccomp library is required as it has the new system call numbers, so as it looks like it will be available early next week will wait for that.

**Comment by voltagex** [comment]
@justincormack epic debugging work there!

I can't see anything new in https://github.com/seccomp/libseccomp and https://tracker.debian.org/pkg/libseccomp says 2.2.3 is going to testing. I'll probably need to raise a bug against seccomp in Debian for it to be updated again if there's a new release.

**Comment by justincormack** [comment]
Ok, libseccomp 2.3.0 has just been released

https://github.com/seccomp/libseccomp/releases/tag/v2.3.0

However the golang bindings have not been updating to match yet, will look to see what is needed there.

**Comment by justincormack** [comment]
I have checked, and simply replacing the `libseccomp` 2.2.3 with 2.3.0 fixes the issue on Debian unstable. I have also tested with updating the docker static build to 2.3.0 and that works too, and I am about to upstream a patch to make that the default.

There is no sign of an update for Debian, may be worth raising an issue there.

**Comment by justincormack** [comment]
PR for statically linked build in #21123

**Comment by voltagex** [comment]
Ping @kees, although I really should just raise a bug via https://bugs.debian.org/cgi-bin/pkgreport.cgi?repeatmerged=no&src=libseccomp - I think the easiest way is to use `reportbug` on a debian system to raise a wishlist bug in the `libseccomp` package. The `debian/watch` file has failed somehow as well, otherwise there'd be a note on https://tracker.debian.org/pkg/libseccomp about there being a new version.

**Comment by justincormack** [comment]
Debian tracker now notes 2.3.0 is available as priority high. https://tracker.debian.org/pkg/libseccomp

**Comment by justincormack** [comment]
Should be fixed in 1.11.2 and 1.12.0 via #22922 although Debian has already upgraded libseccomp so not an issue there now anyway.
